Add KEP #257: Leader Only SubGroup

Edwinhr716 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind documentation
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it

#### Which issue(s) this PR fixes
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of #257 

#### Special notes for your reviewer

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note

```

kerthcet commented:
I may review this tomorrow because of limited bandwidth, sorry for that. 
/assign

kerthcet commented:
reviewing now.

kerthcet commented on keps/257-Subgroup-leader-only/README.md:
```diff
@@ -0,0 +1,353 @@
+
+* `leaderworkerset.sigs.k8s.io/subgroup-policy-type`
+
+In order to keep backwards compatability, it will only be added if the type is `LeaderOnly`.
```
I think even we add the annotation for both types is still backward compatibility. No behavior change.

Edwinhr716 replied:
The scenario that I see is that there is an existing deployment using `SubGroupPolicy` which then upgrades the LWS controller. After the upgrade, it will inject the annotation on the PodSpec, which will then trigger a rolling update since the Pod has changed.

kerthcet commented on keps/257-Subgroup-leader-only/README.md:
````diff
@@ -0,0 +1,353 @@
+}
+```
+
+Because the leader now occupies a subgroup, and `SubGroupSize` will not always be one, the way we calculate the SubGroupIndex needs to be modified. When generating 
````
One argue here is should we mark the leader a sub group? What's the Cons if we don't do this.

Edwinhr716 replied:
That's a good point, hadn't considered not making the leader part of a subgroup at all. Doing it this way simplifies the rest of the changes (though the TPU environment injection will still need some adjustment). Will test it out and add to the `Alternatives` which ever version we don't implement

kerthcet reviewed (commented):
several comments.

Edwinhr716 commented on keps/257-Subgroup-leader-only/README.md:
```diff
@@ -0,0 +1,353 @@
+to the number of workers (so size - 1). 
+
+### Non-Goals
+This KEP assumes that the leader will not request TPU resources, and thus, subGroupSize will always be assumed to be an odd number. Moreover, no TPU environment variables will be 
```
Both TPU and GPU are supported for exclusive placement. I'll rephrase to be accelerator agnostic

Edwinhr716 commented on keps/257-Subgroup-leader-only/README.md:
```diff
@@ -0,0 +1,353 @@
+const (
+	SubGroupPolicyLeaderWorker SubGroupPolicyType = "LeaderWorker"
+
+	SubGroupPolicyLeaderOnly SubGroupPolicyType = "LeaderOnly"
```
I'm fine with `LeaderAlone`, but `All` sounds a bit ambiguous to me. `LeaderAlone` vs `LeaderWorker` maybe? Or something that specifies that the leader and a worker pod will be part of the same subgroup

Edwinhr716 commented:
Changed implementation to the leader not actually being part of any subgroup since it is simpler, added original implementation to Alternatives. Also changed LeaderOnly to LeaderAlone. Please take another look @kerthcet

ahg-g commented on keps/257-Subgroup-leader-only/README.md:
```diff
@@ -0,0 +1,356 @@
+to the number of workers (so size - 1). 
+
+### Non-Goals
+This KEP assumes that the leader will not request TPU resources, and thus, subGroupSize will always be assumed to be an odd number. Moreover, no TPU environment variables will be 
```
pls rephrase to be accelerator agnostic, this applies to both

ahg-g commented on keps/257-Subgroup-leader-only/README.md:
```diff
@@ -0,0 +1,356 @@
+-->
+
+#### Story 1
+As a user, I should be able to create a subGroup that only includes the leader in order to be able to schedule the leader on a different topology while 
```
```suggestion
As a user, I should be able to create a subGroup that only includes the leader in order to be able to schedule the leader with different resource requirements
```

ahg-g commented on keps/257-Subgroup-leader-only/README.md:
````diff
@@ -0,0 +1,356 @@
+```golang
+type SubGroupPolicy struct {
+
+	// +kubebuilder:validation:Enum={LeaderWorker,LeaderAlone}
+	// +kubebuilder:default=LeaderWorker
+	Type SubGroupPolicyType `json:"subGroupPolicyType,omitempty"`
````
fix indentation pls

ahg-g commented on keps/257-Subgroup-leader-only/README.md:
```diff
@@ -0,0 +1,356 @@
+In order to keep backwards compatability, it will only be added if the type is `LeaderAlone`.
+
+### Subgroup Creation 
+Implementation wise, the only change needed is to not add the SubGroup labels on the leader if the SubGroupType is LeaderAlone. Effectively, this means 
```
The indexing within the group will be different now. Currently we don't define an index within the group, we do it only for tpuWorkerId, and this should be calculated differently now to exclude the leader: https://github.com/kubernetes-sigs/lws/blob/main/pkg/utils/accelerators/tpu.go#L113, right?

Edwinhr716 replied:
We already do this when the leader doesn't request TPU resources https://github.com/kubernetes-sigs/lws/blob/main/pkg/utils/accelerators/tpu.go#L115-L117

kerthcet commented on keps/257-Subgroup-leader-only/README.md:
```diff
@@ -0,0 +1,356 @@
+
+### Subgroup Creation 
+Implementation wise, the only change needed is to not add the SubGroup labels on the leader if the SubGroupType is LeaderAlone. Effectively, this means 
+that the leader is not part of a subgroup at all. The only point of a pod being in a subgroup is to guarantee exclusive placement with the other pods 
```
> the leader is not part of a subgroup at all

LeaderAlone is based on the original assumption that leader pod is a separate group, if not, maybe we should rename it to like `leaderExcluded`, in the future, leaderAlone can still be reuse if leader pod should be the 0-index subgroup.

Edwinhr716 replied:
Makes sense to me, will change it to `leaderExcluded`. Does `leaderWorker` then sound good to you for the default setup?

kerthcet replied:
LGTM

kerthcet reviewed (commented):
I think we can leave more implementation details to the PR.

kerthcet commented on keps/257-Subgroup-leader-only/README.md:
```diff
@@ -0,0 +1,355 @@
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+This KEP aims to extend the SubGroup feature by adding an API field that defines the type of Subgroup that will be created. The existing version
+will become "LeaderWorker", and a new type will be added to create a SubGroup that only includes the leader.
```
There's no subgroup for leader pod now.

Edwinhr716 replied:
Right, updated this and all other references to leader being its own subgroup

kerthcet commented on keps/257-Subgroup-leader-only/README.md:
```diff
@@ -0,0 +1,355 @@
+-->
+
+There are cases where the resource requirements for the leader are different from the ones by the worker, while still being able to guarantee exclusive placement
+on the workers. Having a leaderOnly subgroup means that the placement of the leader will be independent of the workers, and by adding the `subgroup-exclusive-topology`, the 
```
There's no leaderOnly type as well.

kerthcet commented on keps/257-Subgroup-leader-only/README.md:
```diff
@@ -0,0 +1,355 @@
+know that this has succeeded?
+-->
+
+Add an option to create a LeaderOnly subgroup, while still being able to create subgroups on the other workers. Therefore, it should be possible to do the following split:
```
ditto

kerthcet reviewed (commented):
I think the calculation of worker subGroup index still needs to change, for example, we set the subgroupSize=2,
- before with group size = 6, it looks like (0, 0), (1, 0), (2, 1), (3, 1), (4, 2), (5 ,2)
- after with group size = 7, it looks like (0, x), (1, 0), (2, 0), (3, 1), (4, 1), (5, 2), (6, 2) 

(x, y) represents the (workerIndex, subGroupIndex), for the workerIndex=2, before the subgroupIndex is 1, after is 0. Not familiar with the TPU implementation, maybe the same? But this is just code implementation, we can leave the detailed review to the PR, just mention this fact in the proposal briefly.

kerthcet commented:
Yes, the result is as we expected, I just mean the method may need to change: https://github.com/kubernetes-sigs/lws/blob/ab40b6285254c4fae69e605578d95f456b64e09f/pkg/webhooks/pod_webhook.go#L245-L251

For workIndex=2, before it's (2, 1), now it's (2, 0), the subGroupIndex changes. Do I miss anything? I think we'll call the method, right?

Edwinhr716 commented:
Yeah we call this method. I guess I'm a little bit confused on what scenario you are talking about. The current behavior when using this feature has workerIndex=2 at subgroup 0. If using the default configuration, but still having an odd number for size, it will still be workerIndex=2 at subgroup 0. The only case where workerIndex=2 gives subgroup 1, is if using default behavior and size is even
